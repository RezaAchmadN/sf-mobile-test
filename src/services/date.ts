export const getCurrentDate = (
    datePlus: number,
    monthPlus: number,
    yearPlus: number
  ) => {
    const date = new Date().getDate() + datePlus;
    const month = new Date().getMonth() + 1 + monthPlus;
    const year = new Date().getFullYear() + yearPlus;
    const dd = (date < 10 ? "0" : "") + date;
    const mm = (month < 10 ? "0" : "") + month;
    return year + "-" + mm + "-" + dd;
  };