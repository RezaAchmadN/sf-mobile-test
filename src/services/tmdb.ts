import { gql } from "@apollo/client";

const TMDBAPIKEY = "48367485a90367721f562c3532360bb3";

export const MOVIE_LIST = gql`
    query Movies {
      Movies(page: $page)
        @rest(
          type: "Movies"
          method: "GET"
          path: "discover/movie?page={args.page}&api_key=${TMDBAPIKEY}"
        ) {
        page 
        total_pages
        results {
          id
          title
          poster_path
          overview
          release_date
          vote_average
        }
      }
    }
  `;

export const POPULAR_MOVIE_LIST = gql`
    query Movies {
      Movies(order: $order, page: $page)
        @rest(
          type: "Movies"
          method: "GET"
          path: "discover/movie?sort_by=popularity.{args.order}&page={args.page}&api_key=${TMDBAPIKEY}"
        ) {
        page 
        total_pages
        results {
          id
          title
          poster_path
          overview
          release_date
          vote_average
        }
      }
    }
  `;

export const NOW_PLAYING_MOVIE_LIST = gql`
  query Movies {
    Movies(gte: $gte, lte: $lte, page: $page)
      @rest(
        type: "Movies"
        method: "GET"
        path: "discover/movie?primary_release_date.gte={args.gte}&primary_release_date.lte={args.lte}&page={args.page}&api_key=${TMDBAPIKEY}"
      ) {
      page 
      total_pages
      results {
        id
        title
        poster_path
        overview
        release_date
        vote_average
      }
    }
  }
`;

export const SEARCH_MOVIE = gql`
query Movies {
  Movies(searchQuery: $searchQuery, page: $page)
    @rest(
      type: "Movies"
      method: "GET"
      path: "search/movie?api_key=${TMDBAPIKEY}&query={args.searchQuery}&page={args.page}"
    ) {
    page
    total_pages
    results @type(name: "Movie") {
      id
      title
      overview
      poster_path
      release_date
      vote_average
    }
  }
}
`;
