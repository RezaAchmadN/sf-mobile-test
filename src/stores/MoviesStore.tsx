import React from "react";

export const MoviesStore = React.createContext<any>([]);

function reducer(movies: [], Action: []): [] {
  return [...movies, ...Action];
}

export function MoviesProvider({
  children,
}: JSX.ElementChildrenAttribute): JSX.Element {
  const [movies, moviesDispatch] = React.useReducer(reducer, []);
  return (
    <MoviesStore.Provider value={{ movies, moviesDispatch }}>
      {children}
    </MoviesStore.Provider>
  );
}
