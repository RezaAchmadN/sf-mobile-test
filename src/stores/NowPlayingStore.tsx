import React from "react";

export const NowPlayingStore = React.createContext<any>([]);

function reducer(nowPLaying: [], action: []): [] {
  return [...nowPLaying, ...action];
}
export default function NowPlayingProvider({
  children,
}: JSX.ElementChildrenAttribute): JSX.Element {
  const [nowPlaying, nowPLayingDispatch] = React.useReducer(reducer, []);
  return (
    <NowPlayingStore.Provider value={{ nowPlaying, nowPLayingDispatch }}>
      {children}
    </NowPlayingStore.Provider>
  );
}
