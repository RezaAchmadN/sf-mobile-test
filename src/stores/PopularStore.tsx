import React from "react";

export const PopularStore = React.createContext<any>([]);

function reducer(popular: [], action: []): [] {
  return [...popular, ...action];
}

export default function PopularProvider({
  children,
}: JSX.ElementChildrenAttribute): JSX.Element {
  const [popular, popularDispatch] = React.useReducer(reducer, []);
  return (
    <PopularStore.Provider value={{ popular, popularDispatch }}>
      {children}
    </PopularStore.Provider>
  );
}
