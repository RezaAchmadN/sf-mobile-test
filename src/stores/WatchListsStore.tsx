import React from "react";
import AsyncStorage from "@react-native-async-storage/async-storage";
import { IAction } from "../interfaces";

const storeData = async (value: any) => {
  try {
    await AsyncStorage.setItem("WatchLists", value);
  } catch (e) {
    console.log(e);
  }
};

const getData = async () => {
  try {
    const value = await AsyncStorage.getItem("WatchLists");
    return JSON.parse(value || "[]");
  } catch (e) {
    console.log(e);
  }
};

export var WatchListsStore = React.createContext<any>([]);

function reducer(movies: [], action: IAction) {
  switch (action.type) {
    case "ADD_FAV":
      const res = [...movies, action.payload];
      storeData(JSON.stringify(res));
      return res;
    case "REMOVE_FAV":
      storeData(JSON.stringify(action.payload));
      return action.payload;
    case "INITIATE":
      return action.payload;
    default:
      return movies;
  }
}

export function WatchListProvider({
  children,
}: JSX.ElementChildrenAttribute): JSX.Element {
  const [watchLists, watchListsDispatch] = React.useReducer(reducer, []);

  React.useEffect(() => {
    getData().then((value: any) =>
      watchListsDispatch({ type: "INITIATE", payload: value })
    );
  }, []);

  return (
    <WatchListsStore.Provider value={{ watchLists, watchListsDispatch }}>
      {children}
    </WatchListsStore.Provider>
  );
}
