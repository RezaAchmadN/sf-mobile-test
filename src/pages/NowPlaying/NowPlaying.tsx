import React, { useEffect, useState } from "react";
import {
  ActivityIndicator,
  FlatList,
  SafeAreaView,
  TouchableOpacity,
} from "react-native";
import { Button } from "react-native-elements";
import Feather from "react-native-vector-icons/Feather";

import { Styles } from "../../../Styles";
import MovieCard from "../../components/MovieCard";
import { getCurrentDate } from "../../services/date";
import { fetchNowPlayingMovies } from "../../Actions";

export default function NowPlaying({ navigation }: any) {
  const [Movies, setMovies] = useState<any>([]);
  const [Page, setPage] = useState(1);
  const [TotalPage, setTotalPage] = useState<number>(1);
  const [Loading, setLoading] = useState<boolean>(false);

  useEffect(() => {
    setLoading(true);
    fetchNowPlayingMovies(
      Page,
      getCurrentDate(0, -1, 0),
      getCurrentDate(0, 0, 0)
    )
      .then((res: any) => {
        setMovies([...Movies, ...res.data.results]);
        setPage(res.data.page);
        setTotalPage(res.data.total_pages);
      })
      .finally(() => setLoading(false));
  }, []);

  React.useLayoutEffect(() => {
    navigation.setOptions({
      headerRight: () => (
        <Button
          onPress={() => navigation.push("Search")}
          icon={<Feather name="search" size={30} />}
          type="clear"
        />
      ),
    });
  }, [navigation]);

  return (
    <SafeAreaView>
      <FlatList
        style={Styles.movieLayout}
        contentContainerStyle={{ paddingBottom: 32 }}
        data={Movies}
        keyExtractor={(item, index) => index.toString()}
        onEndReachedThreshold={0.1}
        onEndReached={() => {
          setLoading(true);
          if (TotalPage > Page)
            fetchNowPlayingMovies(
              Page + 1,
              getCurrentDate(0, -1, 0),
              getCurrentDate(0, 0, 0)
            )
              .then((res: any) => {
                setMovies([...Movies, ...res.data.results]);
                setPage(res.data.page);
                setTotalPage(res.data.total_pages);
              })
              .finally(() => setLoading(false));
        }}
        renderItem={({ item }) => (
          <TouchableOpacity
            onPress={() =>
              navigation?.push("MovieDetail", {
                title: item.title,
                props: item,
              })
            }
          >
            <MovieCard {...item} />
          </TouchableOpacity>
        )}
      />
      {Loading && (
        <ActivityIndicator
          style={Styles.center}
          size="large"
          color={"#0000ff"}
          testID="loading"
        />
      )}
    </SafeAreaView>
  );
}
