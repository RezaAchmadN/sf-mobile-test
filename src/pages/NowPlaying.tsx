import { useLazyQuery } from "@apollo/client";
import React, { useEffect, useState } from "react";
import {
  ActivityIndicator,
  FlatList,
  SafeAreaView,
  Text,
  TouchableOpacity,
  View,
} from "react-native";
import { Button } from "react-native-elements";
import Feather from "react-native-vector-icons/Feather";

import { Styles } from "../../Styles";
import { NOW_PLAYING_MOVIE_LIST } from "../services/tmdb";

const MovieList = React.lazy(() => import("../components/MovieCard"));

const getCurrentDate = (
  datePlus: number,
  monthPlus: number,
  yearPlus: number
) => {
  const date = new Date().getDate() + datePlus;
  const month = new Date().getMonth() + 1 + monthPlus;
  const year = new Date().getFullYear() + yearPlus;
  const dd = (date < 10 ? "0" : "") + date;
  const mm = (month < 10 ? "0" : "") + month;
  return year + "-" + mm + "-" + dd;
};

export default function NowPlaying({ navigation }: any) {
  const [Movies, setMovies] = useState<any>([]);
  const [Page, setPage] = useState(1);
  const [TotalPage, setTotalPage] = useState<number>(1);

  const [fetchMovieList, { loading, error }] = useLazyQuery(
    NOW_PLAYING_MOVIE_LIST,
    {
      onCompleted: (data) => {
        setMovies([...Movies, ...data.Movies.results]);
        setPage(data.Movies.page);
        setTotalPage(data.Movies.total_pages);
      },
    }
  );

  useEffect(() => {
    fetchMovieList({
      variables: {
        gte: getCurrentDate(0, -1, 0),
        lte: getCurrentDate(0, 0, 0),
        page: Page,
      },
    });
  }, []);

  React.useLayoutEffect(() => {
    navigation.setOptions({
      headerRight: () => (
        <Button
          onPress={() => navigation.push("Search")}
          icon={<Feather name="search" size={30} />}
          type="clear"
        />
      ),
    });
  }, [navigation]);

  return (
    <SafeAreaView>
      {error && <Text>`Error! ${error}`</Text>}
      <React.Suspense
        fallback={
          <View style={Styles.center}>
            <ActivityIndicator size="large" color={"#0000ff"} />
          </View>
        }
      >
        <FlatList
          style={Styles.movieLayout}
          contentContainerStyle={{ paddingBottom: 32 }}
          data={Movies}
          keyExtractor={(item, index) => index.toString()}
          onEndReachedThreshold={0.1}
          onEndReached={() => {
            if (TotalPage > Page)
              fetchMovieList({
                variables: {
                  gte: getCurrentDate(0, -1, 0),
                  lte: getCurrentDate(0, 0, 0),
                  page: Page + 1,
                },
              });
          }}
          renderItem={({ item }) => (
            <TouchableOpacity
              onPress={() =>
                navigation?.push("MovieDetail", { title: item.title })
              }
            >
              <MovieList {...item} />
            </TouchableOpacity>
          )}
        />
      </React.Suspense>
      {loading && (
        <ActivityIndicator
          style={Styles.center}
          size="large"
          color={"#0000ff"}
        />
      )}
    </SafeAreaView>
  );
}
