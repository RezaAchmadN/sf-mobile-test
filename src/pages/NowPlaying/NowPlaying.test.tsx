import React from "react";
import { act, cleanup, fireEvent, render } from "@testing-library/react-native";
import { MockedProvider } from "@apollo/client/testing";
import { MOVIE_LIST, NOW_PLAYING_MOVIE_LIST } from "../../services/tmdb";
import NowPlaying from "./NowPlaying";
import { WatchListProvider } from "../../stores/WatchListsStore";
import { getCurrentDate } from "../../services/date";

const props = {
  navigation: {
    navigate: jest.fn(),
    setOptions: jest.fn(),
  },
} as any;

const mocks = [
  {
    request: {
      query: NOW_PLAYING_MOVIE_LIST,
      variables: {
        gte: getCurrentDate(0, -1, 0),
        lte: getCurrentDate(0, 0, 0),
        page: 1,
      },
    },
    result: {
      data: {
        Movies: {
          __typename: "Movies",
          page: 1,
          results: [
            {
              __typename: "Movie",
              id: 603,
              title: "The Matrix",
              overview:
                "Set in the 22nd century, The Matrix tells the story of a computer hacker who joins a group of underground insurgents fighting the vast and powerful computers who now rule the earth.",
              poster_path: "/dXNAPwY7VrqMAo51EKhhCJfaGb5.jpg",
              release_date: "1999-03-30",
              vote_average: 8.1,
            },
            {
              __typename: "Movie",
              id: 604,
              title: "The Matrix Reloaded",
              overview:
                "Six months after the events depicted in The Matrix, Neo has proved to be a good omen for the free humans, as more and more humans are being freed from the matrix and brought to Zion, the one and only stronghold of the Resistance.  Neo himself has discovered his superpowers including super speed, ability to see the codes of the things inside the matrix and a certain degree of pre-cognition. But a nasty piece of news hits the human resistance: 250,000 machine sentinels are digging to Zion and would reach them in 72 hours. As Zion prepares for the ultimate war, Neo, Morpheus and Trinity are advised by the Oracle to find the Keymaker who would help them reach the Source.  Meanwhile Neo's recurrent dreams depicting Trinity's death have got him worried and as if it was not enough, Agent Smith has somehow escaped deletion, has become more powerful than before and has fixed Neo as his next target.",
              poster_path: "/jBegA6V243J6HUnpcOILsRvBnGb.jpg",
              release_date: "2003-05-15",
              vote_average: 6.9,
            },
          ],
          total_pages: 2,
        },
      },
    },
  },
  {
    request: {
      query: NOW_PLAYING_MOVIE_LIST,
      variables: {
        gte: getCurrentDate(0, -1, 0),
        lte: getCurrentDate(0, 0, 0),
        page: 2,
      },
    },
    result: {
      data: {
        Movies: {
          __typename: "Movies",
          page: 2,
          results: [
            {
              __typename: "Movie",
              id: 605,
              title: "The Matrix Revolutions",
              overview:
                "The human city of Zion defends itself against the massive invasion of the machines as Neo fights to end the war at another front while also opposing the rogue Agent Smith.",
              poster_path: "/fgm8OZ7o4G1G1I9EeGcb85Noe6L.jpg",
              release_date: "2003-11-03",
              vote_average: 6.6,
            },
          ],
          total_pages: 2,
        },
      },
    },
  },
];

const networkErrorMocks = [
  {
    request: {
      query: NOW_PLAYING_MOVIE_LIST,
      variables: { page: 1 },
    },
    error: new Error("An error occurred"),
  },
];

describe("NowPlaying", () => {
  afterEach(cleanup);

  it("Testing the 'loading'", () => {
    const component = render(
      <MockedProvider mocks={[]} addTypename={false}>
        <NowPlaying {...props} />
      </MockedProvider>
    );

    expect(component.getByTestId("loading")).toBeTruthy;
  });

  it("on Success", async () => {
    const component = render(
      <MockedProvider mocks={mocks} addTypename={false}>
        <WatchListProvider>
          <NowPlaying {...props} />
        </WatchListProvider>
      </MockedProvider>
    );

    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0));
    });
    expect(component.getAllByTestId("MovieCard").length).toBe(2);
  });

  it("on Network Error", async () => {
    const component = render(
      <MockedProvider mocks={networkErrorMocks} addTypename={false}>
        <WatchListProvider>
          <NowPlaying {...props} />
        </WatchListProvider>
      </MockedProvider>
    );

    await act(async () => {
      await new Promise((resolve) => setTimeout(resolve, 0));
    });
    expect(component.getByText("Error!")).toBeTruthy;
  });

  // it("onClicked", async () => {
  //   const component = render(
  //     <MockedProvider mocks={mocks} addTypename={false}>
  //       <WatchListProvider>
  //         <Home {...props} />
  //       </WatchListProvider>
  //     </MockedProvider>
  //   );

  //   await act(async () => {
  //     await new Promise((resolve) => setTimeout(resolve, 0));
  //   });
  //   component.debug()
  //   await fireEvent.press(await component.getByText("The Matrix Reloaded"))
  //   expect(component.getByText("Error!")).toBeTruthy;
  // });

  // it("on Scrolled", async () => {
  //   const component = render(
  //     <MockedProvider mocks={mocks} addTypename={false}>
  //       <WatchListProvider>
  //         <Home {...props} />
  //       </WatchListProvider>
  //     </MockedProvider>
  //   );

  //   const eventData = {
  //     nativeEvent: {
  //       contentOffset: {
  //         y: 500,
  //       },
  //       contentSize: {
  //         // Dimensions of the scrollable content
  //         height: 500,
  //         width: 100,
  //       },
  //       layoutMeasurement: {
  //         // Dimensions of the device
  //         height: 100,
  //         width: 100,
  //       },
  //     },
  //   };

  //   await act(async () => {
  //     await new Promise((resolve) => setTimeout(resolve, 0));
  //   });
  //   (await component.findByTestId("FlatList")).instance.end;

  //   const onEndReached = jest.fn();
  //   await component.debug()
  // });
});
