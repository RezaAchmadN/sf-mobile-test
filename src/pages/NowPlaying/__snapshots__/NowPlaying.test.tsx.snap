// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Correct API Match Snapshot 1`] = `[Function]`;
